### DISCUSSION TITLE
undocumented(?) features when setting hs-colors on xy-color-"only"-device

### DISCUSSION THREAD

**Discussion by GsakuL** [Koenkk/zigbee2mqtt#7931] [General]
I'm using the Z2M-Docker with version '1.20.0 (commit #41b67fd)'; and FHEM as my MQTT2/SmartHome server.
I was setting up RGB controller [Gledopto (GL-C-008-1ID)](https://www.zigbee2mqtt.io/devices/GL-C-008-1ID.html#gledopto-gl-c-008-1id) and some bulbs [(Müller Licht / Tint 404000/404005/404012)](https://www.zigbee2mqtt.io/devices/404000_404005_404012.html), when I noticed that Alexa (at least via FHEMs 'Alexa Connector') seems to work only with Hue(+Saturation) instead of RGB. The bulbs support HS by themselves, so i just set up appropiate "setList-commands" in FHEM.
`hue:slider,0,1,360 $DEVICETOPIC/set {"state":"on","color":{"$EVTPART0":"$EVTPART1"}}` which could result in `{"state":"on","color":{"hue":"140"}}`. It worked, and I moved on to the controller.
First I used some crude conversion functions (Hue=>RGB) within FHEM to set the desired colors. This worked as a PoC.
Later I copied the same "pass through"-commands from the bulbs to the GLEDOPTO device and it just worked. Which is nice. But it took me a long time to get there.

Why does the documentation for the GLEDOPTO (and probably other xy-only devices) say it's possible to set color via XY and *alternatively* via RGB/HEX. Sure it doesn't say say you can't via HS, but I have assumed as much. Is this some internal conversion in Z2M? The payload seems to suggest that.
Is there some general doc topic that I missed?
I like this very much, as it reduces clutter and/or complexity in FHEM. I just wish I found this sooner.

**Comment by Koenkk** [comment]
Added this now for Gledopto devices, this has to be explicitly enabled (since hue/saturation is optional, and color x/y is mandator according to the zigbee spec).

Docs will be updated with next release.

**Comment by GsakuL** [comment]
just further FYI's. *I know this may not belong here, but it's for context/completion. (If, maybe some stumbles upon this later)* 

I just got some [TuYa TS0503B](https://www.zigbee2mqtt.io/devices/TS0503B.html) RGB Controllers in the mail. And they also respond to HS.

One note regarding FHEM (and possibly other servers?): Especially when setting up the device(s), it may be beneficial to send both hue **and** saturation for more consistent results. This can be achived e.g. using these `setList`-Lines:
```
hue:colorpicker,HUE,0,1,360 { "$DEVICETOPIC/set " . constructHS($EVTPART1, ReadingsNum($NAME, "color_saturation",100));}
saturation:slider,0,1,100 { "$DEVICETOPIC/set " . constructHS(ReadingsNum($NAME, "color_hue",0), $EVTPART1);}
```
and this coresponding function in 99_myUtils.pm:
```pm
sub constructHS($$)
{
  my ($h, $s) = @_;
  return '{"state":"on","color_mode":"hs","color":{"hue":' . int($h). ', "saturation": '. int($s) .'}}';
}
```
It uses the last known value for s/h respectively. My setup thus needs no color-conversion functions/"readings".
